Raw Lab colorspace access (for higher performance)

benjie opened the pull request:
This PR builds on #13 and enables operating in the raw Lab colorspace, increasing performance by a further 200% if you're willing to handle your own RGB conversion.

```
rnd.gen 2.9617803728544567 ± 0.1637196912746409
colorDiff.closest 155.74627221216926 ± 6.77161445118158
colorDiff.closest [predef] 144.12553840911875 ± 3.7143865877309605
lib.palette_lab [predef] 47.268267507105605 ± 2.2698601609636198
```

<details>
<summary>Benchmark code</summary>

```js
const Benchmark = require('benchmark');
const colorDiff = require('./lib');

const suite = new Benchmark.Suite;

const rnd = function () {
  return {
    R: Math.floor(Math.random() * 256),
    G: Math.floor(Math.random() * 256),
    B: Math.floor(Math.random() * 256),
  };
};

const rndA = function () {
  return [
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
    rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(), rnd(),
  ];
};

const predefTarget = rnd();
const predefPalette = rndA();
const predefTargetLab = colorDiff.rgb_to_lab(predefTarget);
const predefPaletteLab = predefPalette.map(colorDiff.rgb_to_lab);

suite
  .add('rnd.gen', function () {
    return [rnd(), rndA()];
  })
  .add('colorDiff.closest', function () {
    return colorDiff.closest(rnd(), rndA());
  })
  .add('colorDiff.closest [predef]', function () {
    return colorDiff.closest(predefTarget, predefPalette);
  })
  .add('lib.palette_lab [predef]', function () {
    return colorDiff.palette_lab(predefTargetLab, predefPaletteLab, true);
  })
  .on('complete', function () {
    for (let i = 0; i < 4; i += 1) {
      console.log(this[i].name, this[i].stats.mean * 1000000, '±', this[i].stats.deviation * 1000000);
    }
  })
  .run({});
```
</details>

~~Tests to follow...~~

benjie commented:
When reviewing, you're best looking at the 4th commit only (https://github.com/markusn/color-diff/pull/14/commits/52c2972a7676367543518b13fc8d3b55935faadf), since the first 3 are in #13 already, and the 5th is just tests.

benjie commented:
Here's a profile of the benchmark running, `ciede2000` is (not unexpectedly) where the majority of the time is spent, interestingly `pow` uses a LOT of CPU time!

![screenshot 2017-04-14 07 45 59](https://cloud.githubusercontent.com/assets/129910/25035611/760083b6-20e6-11e7-85d1-ac91e7e05704.png)

benjie commented:
I have rebased this onto master

markusn commented on lib/palette.js:
```diff
@@ -98,6 +99,35 @@ function map_palette(a, b, type, bc)
+* @param 'findFurthest' should be falsy to find the closest color
+* @return {labcolor}
+*/
+function palette_lab(targetColor, palette, findFurthest)
```
I would prefer if this was implemented like map_palette (a map_lab_palette) instead and then implement the closest/furthest function as it is implemented for RGB(A) in index.js. This should not give much performance overhead and make it possible to use for mapping multiple colors, which is useful. Preferably the function arguments should be similar to map_palette (minus the background color)

benjie replied:
I originally implemented it in that way, but the jitter was much higher, I assume because of the object allocation (`c`) and the string construction (`palette_map_to_key`) was causing GC overhead.

Would you be happy if I made another method that implemented this behaviour on top of the current implementation, i.e.:

```js
function map_palette_lab(a, b, find_furthest) {
  return b.reduce(function(memo, color) {
    memo[lab_map_key(color)] = palette_lab(color, b, find_furthest);
    return memo;
  }, {});
}
```

markusn commented on test/palette.js:
```diff
@@ -42,9 +43,20 @@ var navy    = {'R':0   , 'G':0   ,'B':128};
 var yellow  = {'R':255 , 'G':255 ,'B':0};
 var gold    = {'R':255 , 'G':215 ,'B':0};
+
+var whiteLab   = color_convert.rgb_to_lab(white);
```
A general note: use snake_case instead of camelCase as this is what is used for the rest of the code.

benjie replied:
Oops, sorry!

benjie commented:
- Fixed the case of variables from camelCase to snake_case
- Added `map_palette_lab` method to match `map_palette` method
- Added tests for map_palette_lab
- Added additional exports `closest_lab` and `furthest_lab`
- Rebased for a tidy history

markusn commented on lib/index.js:
```diff
@@ -27,3 +30,17 @@ color.furthest = function(target, relative, bc) {
+
+color.furthest_lab = function(target, relative) {
+    var key = color.lab_palette_map_key(target);
+    var result = color.map_palette_lab([target], relative, 'furthest');
```
Why not use match directly here? From your previous comment it should yield slightly better performance. Same for closest_lab.

benjie replied:
Good point.

markusn commented on test/palette.js:
```diff
@@ -112,6 +132,89 @@ describe('palette', function(){
+  describe('#map_palette_lab()', function (){
+    it('should map all colors to themselves when possible #1',
+       function(){
+         var expected1_1                                    = {};
```
Align =.  Same further down in the other new tests.

benjie replied:
👍

benjie commented:
Done 👍

benjie commented:
🎊
